aten/src/ATen/Allocator.h

```diff
@@ -0,0 +1,33 @@
+#pragma once
+
+#include <memory>
+#include <stddef.h>
+
+#include "ATen/Retainable.h"
+
+namespace at {
+
+struct Allocator {
+  virtual void* allocate(std::size_t n) const = 0;
+  virtual void deallocate(void* ptr) const = 0;
+};
+
+namespace detail {
+
+struct AllocatorContext : public Retainable {
```

Comment:
I’d say it’s a bit weird that AllocatorContext owns an Allocator (would be more natural the other way around). Not sure how would one implement a caching allocator if it could be owned only by a single context. Any reason to make it this way?

Reply:
Allocators own their own state, if they have any. The state is not part of `AllocatorContext`. 

`AllocatorContext` is an implementation detail (it's not part of the public API). It adds refcounting to allocators without requiring refcounting to be part of the public API.  If I were designing the API from scratch, I might have `Storage` own the allocator and not bother with `AllocatorContext`:

```
struct Storage {
private:
  std::unique_ptr<Allocator> allocator;
};
```

Instead, we have to thread the allocator through the `THAllocator`/`THCDeviceAllocator` API, which uses C style `void*` context pointers. We want refcounting internally because a resized tensor goes through the following calls:

```
malloc (new)
...
malloc (resize)
free (delete old)
...
free (final delete)
```

We need refcounting to know if we can delete the allocator `unique_ptr` at the final call to free.

You can implement stateful or stateless allocators with this API or allocators that only have global state. The pinned memory allocator is a caching allocator.

For stateless or global allocators, there's a little extra overhead with this API, since it takes in a `unique_ptr<Allocator>` whose lifetime is tied to the storage. This is the same tradeoff we make with `storageFromBlob(..., deleter)` and the same tradeoff made by `shared_ptr` with a custom deleter.

Reply:
Having the storage own a single copy of an allocator is even more surprising for me. How do you implement CUDA caching allocator? Do you just make the instances simple objects with very little state, and use static members?

Reply:
Yes. As I wrote above, this is the same ownership model used by `storageFromBlob` with a deleter and the `stl`'s shared_ptr. The lifespan of the allocator needs to be at least as long as the lifetime of the storage.

Reply:
I guess my problem is that I don’t understand where the Context suffix is coming from, since this object doesn’t hold any extra information that an allocator could use. Why not use the pImpl pattern that’s everywhere and call the refcounted thing an Allocator?

Reply:
We could just make `Allocator : public Retainable` and add refcounting to every allocator. I didn't do this because I didn't want refcounting to be part of the `Allocator` interface. Refcounting is an implementation detail here: it's only necessary because of the mismatch between `ATen` and the `THAllocator` API.

Reply:
Maybe, but when you see `Context`, you assume that this holds some useful piece of information, while this really is `std::shared_ptr<Allocator>`.

Reply:
I don't care too much about the name, especially since it's hidden away in the `detail` namespace. I called it context because it's the value of the `void* ctx` in the `THAllocator.

Reply:
I agree the Context suffix isn't great; since it's not part of the public API it seems fine to leak the implementation details and just call it an AllocatorRetainable (switching the order is probably clearer but it's nice to have Allocator at the front).

Reply:
Six months later, I wonder: why can't there just be one static lifetime copy of the pinned memory allocator?